src/config.rs

```diff
@@ -0,0 +1,275 @@
+use serde::{Deserialize, Serialize};
+use std::error;
+use std::io::Write;
+use uuid::Uuid;
+
+use tiny_http::{Response, Server};
+
+use std::collections::HashMap;
+use url::Url;
+
+use serde_json::Value;
+use tungstenite::{connect, Message};
+
+use ha_api::types::{GetAccessTokenResponse, RegisterDeviceResponse};
+use ha_api::HomeAssistantAPI;
+
+// HA creates tokens for 10 years so we do the same
+const LONG_LIVED_TOKEN_VALID_FOR: u32 = 365;
+
+const LONG_LIVED_TOKEN_WS_COMMAND_ID: u32 = 11;
+
+const LOCAL_SERVER_HOST: &str = "0.0.0.0:8000";
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct HaConfig {
+    pub host: String,
+    #[serde(rename = "long-lived-token")]
+    pub long_lived_token: Option<String>,
+    #[serde(rename = "device-id")]
+    pub device_id: Option<String>,
+    #[serde(rename = "webhook-id")]
+    pub webhook_id: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct YamlConfig {
+    pub ha: HaConfig,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct WsAuthRequest {
+    #[serde(rename(serialize = "type"))]
+    auth_type: String,
+    access_token: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct WsLongLivedAccessTokenRequest {
+    id: u32,
+    #[serde(rename(serialize = "type"))]
+    command_type: String,
+    client_name: String,
+    lifespan: u32,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct WsLongLivedAccessTokenResponse {
+    id: u32,
+    #[serde(rename(deserialize = "type"))]
+    command_type: String,
+    success: bool,
+    result: Option<String>,
+}
+
+type Result<T> = std::result::Result<T, Box<dyn error::Error>>;
+
+pub async fn wait_for_token(
+    config: &YamlConfig,
+    ha_api: &HomeAssistantAPI,
+) -> Result<GetAccessTokenResponse> {
+    let maybe_server = Server::http(LOCAL_SERVER_HOST);
+    match maybe_server {
+        Ok(server) => {
+            println!("Open http://{}/auth/authorize?client_id=http%3A%2F%2Flocalhost%3A8000&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fcallback in your browser", config.ha.host.as_str());
```

Comment:
It'd be cool to use `LOCAL_SERVER_HOST` here, although not sure how much of a pain that'll be?

Reply:
That possible.